R/common-impute.R

```diff
@@ -48,33 +48,95 @@ groupedLOCF <- function(df, id) {
 }
 
 #' @title
-#' Perform imputation on a vector
-#'
-#' @description This class performs imputation on a vector. For numeric vectors
-#' the vector-mean is used; for factor columns, the most frequent value is used.
-#' @param v A vector, or column of values with NAs.
-#' @return A vector, or column of values now with no NAs
+#' Perform imputation on a dataframe
 #'
+#' @description This class performs imputation on a data frame. For numeric columns,
+#' the column-mean is used; for factor columns, the most frequent value is used.
+#' @param df A dataframe of values with NAs.
+#' @param imputeVals A list of values to be used for imputation. Must be the same length
+#' as the number of columns in df.
+#' @return A list. The first element, df, is the imputed dataframe. The second element,
+#' imputeVals, is a list of the imputation value used.
+#' #'
 #' @export
 #' @references \url{http://healthcare.ai}
 #' @seealso \code{\link{healthcareai}}
 #' @examples
-#' # For a numeric vector
-#' vResult <- imputeColumn(c(1,2,3,NA))
-#'
-#' # For a factor vector
-#' vResult <- imputeColumn(c('Y','N','Y',NA))
-#'
-#' # To use this function on an entire data frame:
-#' df <- data.frame(a=c(1,2,3,NA),
-#'                 b=c('Y','N','Y',NA))
-#' df[] <- lapply(df, imputeColumn)
-#' head(df)
-imputeColumn <- function(v) {
-  if (is.numeric(v)) {
-    v[is.na(v)] <- mean(v, na.rm = TRUE)
+#' 
+#' # Impute a single column
+#' df <- data.frame(a=c(1,2,3,NA), b=c('Y','N','Y',NA),
+#'    c=c(11,21,31,43), d=c('Y','N','N',NA))
+#' df <- df['a'] # note df[,1] does not return a df!
```

Comment:
As an aside, this inconsistency is one issue that `tibble` (part of the `tidyverse`) solves. It's especially insidious in data.frames if the column selection is programmatic and you don't know if you'll get back one or many columns. Eg:

```
> class(tibble::data_frame(x = 1:4, y = 2:5)[ , 1] )
[1] "tbl_df"     "tbl"        "data.frame"
> class(data.frame(x = 1:4, y = 2:5)[ , 1] )
[1] "integer"
```